include/ginkgo/core/solver/idr.hpp

```diff
@@ -262,33 +232,18 @@ class Idr : public EnableLinOp<Idr<ValueType>>,
                  std::shared_ptr<const LinOp> system_matrix)
         : EnableLinOp<Idr>(factory->get_executor(),
                            gko::transpose(system_matrix->get_size())),
+          EnablePreconditionedIterativeSolver<ValueType, Idr<ValueType>>{
+              std::move(system_matrix), factory->get_parameters()},
           parameters_{factory->get_parameters()},
-          system_matrix_{std::move(system_matrix)}
-    {
-        if (parameters_.generated_preconditioner) {
-            GKO_ASSERT_EQUAL_DIMENSIONS(parameters_.generated_preconditioner,
-                                        this);
-            set_preconditioner(parameters_.generated_preconditioner);
-        } else if (parameters_.preconditioner) {
-            set_preconditioner(
-                parameters_.preconditioner->generate(system_matrix_));
-        } else {
-            set_preconditioner(matrix::Identity<ValueType>::create(
-                this->get_executor(), this->get_size()[0]));
-        }
-        stop_criterion_factory_ =
-            stop::combine(std::move(parameters_.criteria));
-        subspace_dim_ = parameters_.subspace_dim;
-        kappa_ = parameters_.kappa;
-        deterministic_ = parameters_.deterministic;
-        complex_subspace_ = parameters_.complex_subspace;
-    }
+          subspace_dim_{parameters_.subspace_dim},
+          kappa_{parameters_.kappa},
+          deterministic_{parameters_.deterministic},
+          complex_subspace_{parameters_.complex_subspace}
```

Comment:
I think we talked discussed this before. Given that the parameters struct is public, we can just use that and not have duplicate member variables.